Added arm64 jobs for Travis-CI

odidev opened the pull request:
- Added arm64 jobs in travis.yml
- Added Cibuildwheel to build wheels for ARM64 and x86 platforms.

odidev commented:
@bdarnell,
Wheels can be uploaded to PyPI by using twine. We can follow https://github.com/joerick/cibuildwheel-autopypi-example for adding twine process to travis.
Please let me know your thoughts on the same.
Thanks.

bdarnell commented on .travis.yml:
```diff
@@ -22,10 +22,21 @@ jobs:
+      arch: arm64
+      env: TOX_ENV=py37-full
+    - python: '3.8'
+      service: docker
```
It looks like this should be `services`, not `service` (although it seems to work anyway?)

odidev replied:
Yes, my bad. Corrected the same.

bdarnell commented on .travis.yml:
```diff
@@ -45,6 +62,11 @@ install:
+    - |
+      if [[ -n ${BUILD_WHEEL} ]]; then
+         pip install -U cibuildwheel
+         cibuildwheel --output-dir dist/
```
`cibuildwheel` is relatively slow; we shouldn't run it until the end of `script` instead of here at the end of `install` (Alternatively, we could use `cibuildwheel` as the first step in all cases, and have tox test the built wheel instead of building its own). 

It might be better to have each travis configuration build its own wheel instead of building them all from py38 (to ensure we test what we build, and also to have some parallelism in the CI pipeline). But I'd also like to move to the stable ABI so we'd no longer have per-python-version wheels.

odidev replied:
Added cibuildwheel as first step in each case (as suggested). also installing tornado from it's own build wheel and running test on the same.

odidev commented on .travis.yml:
```diff
@@ -22,10 +22,21 @@ jobs:
       env: TOX_ENV=py35-full
     - python: '3.6'
       env: TOX_ENV=py36-full
+    - python: '3.6'
```
No, there is no specific reason for not running for python3.5, added for py35 as well.

odidev commented on .travis.yml:
```diff
@@ -36,6 +47,12 @@ jobs:
+    - python: '3.8'
+      env: TOX_ENV=lint
+    - python: '3.8'
+      arch: arm64
```
Sure, amended PR to remove "lint" build for arm64 platform.

bdarnell commented on .travis.yml:
```diff
@@ -35,13 +55,23 @@ jobs:
+      if [[ -n ${BUILD_WHEEL} ]]; then
+         pip install -U cibuildwheel
+         cibuildwheel --output-dir dist
+         rm -rf dist/*manylinux2010_x86_64.whl
```
Why is this line necessary? Why is it only on linux and how would I know when/if the 2010 or other parts of the line need to change?

ploxiln replied:
I think this question is still relevant, and now "sudo" has also been added to just this command, raising the same question: why is it needed?

bdarnell commented on .travis.yml:
```diff
@@ -35,13 +55,23 @@ jobs:
+         pip install -U cibuildwheel
+         cibuildwheel --output-dir dist
+         rm -rf dist/*manylinux2010_x86_64.whl
+         pip install dist/*.whl
```
`tox` runs all the tests in an isolated virtualenv, so installing the wheel here doesn't cause it to get tested. I think you need to tell tox to use the wheel that was just built.

odidev commented:
@bdarnell,
Thanks for your valuable suggestions, Here are the following changes:
- Modified "tox.ini", Increased "ASYNC_TEST_TIMEOUT" from 15 to 25 as few tests were getting timeout error while testing.
- Modified ".travis.yml", now running the tests using "CIBW_TEST_COMMAND", which will use the same wheel which has been built using "cibuildwheel"

Since "CIBW_TEST_COMMAND" tests the wheel in a separate docker container, one test is getting failed in amd64 platform, below are the logs:

```
FAIL: test_bind_source_ip (tornado.test.curl_httpclient_test.CurlHTTPClientCommonTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
File "/project/.tox/py38-full/lib/python3.8/site-packages/tornado/testing.py", line 98, in _call_
result = self.orig_method(*args, **kwargs)
File "/project/.tox/py38-full/lib/python3.8/site-packages/tornado/testing.py", line 599, in post_coroutine
return self.io_loop.run_sync(
File "/project/.tox/py38-full/lib/python3.8/site-packages/tornado/ioloop.py", line 530, in run_sync
return future_cell[0].result()
File "/project/.tox/py38-full/lib/python3.8/site-packages/tornado/gen.py", line 748, in run
yielded = self.gen.send(value)
File "/project/.tox/py38-full/lib/python3.8/site-packages/tornado/test/httpclient_test.py", line 560, in test_bind_source_ip
yield self.http_client.fetch(request)
cibuildwheel-ce44eb1b-8c67-4ea2-8467-37ca8e65b549
AssertionError: (<class 'ValueError'>, <class 'tornado.httpclient.HTTPClientError'>) not raised

```
Tested the same directly on amd64 machine (Not is docker container), All the tests were passing there.
Please suggest your thoughts on the same.
Thanks

bdarnell commented on .travis.yml:
```diff
@@ -7,7 +7,13 @@ addons:
+env:
+  global:
+    - CIBW_ENVIRONMENT="PYCURL_SSL_LIBRARY=nss"
+    - CIBW_BEFORE_BUILD="yum install -y libcurl-devel openssl-devel"
```
Why do we install `openssl-devel` but tell pycurl to use `nss`? 

This could go in `CIBW_BEFORE_TEST` instead of `CIBW_BEFORE_BUILD`. Our build itself doesn't need any extra dependencies beyond what CIBW provides.

odidev replied:
Yes, `openssl-devel` is not required here. In fact whole Global env section is not required now, as we are testing the built wheel using `"tox --installpkg"`.

bdarnell commented on .travis.yml:
```diff
@@ -35,16 +61,30 @@ jobs:
+    - |
+      if [[ -n ${BUILD_WHEEL} ]]; then
+         # To test it's own built wheel
+         export CIBW_TEST_COMMAND="cd {project} && tox -e $TOX_ENV"
```
Are you sure this is testing the built wheel? I thought `tox` would do its own build unless you pass `--installpkg` and tell it where the built wheel lives

odidev replied:
I'm quite sure that it was testing the built wheel only, Here is the reference:
[cibuildwheel](https://cibuildwheel.readthedocs.io/en/stable/options/#test-command)
But anyway, Modified the same as you suggested, Now testing the built wheel using `--installpkg` with `tox` command.

bdarnell replied:
That link doesn't really answer the question. Installing the built wheel in the top-level environment doesn't install it in the environments tox creates for the tests themselves. I'm pretty sure `--installpkg` is necessary here.

bdarnell commented on .travis.yml:
```diff
@@ -35,16 +61,30 @@ jobs:
     - python: nightly
 
 install:
+    - |
```
I think this should go in the `script` phase instead of the `install` phase. That puts it after the diagnostic prints (`python -VV`, `curl-config`, etc) and means there's just one `if/else` on `BUILD_WHEEL` instead of an `if -n` one place and an `if -z` in another.

odidev replied:
Correct, Thanks for the suggestion. Modified the same.

bdarnell reviewed (commented):
Hmm, I don't know much about what docker does with networking. It's also possible that this is related to a different version of libcurl used in this container. In any case, this test isn't terribly interesting, especially if it's only failing on `curl_httpclient_test`. I think what I'd do is split up `test_bind_source_ip`, leaving the first half (the successful use of `127.0.0.1`) where it is and moving the other half to `test_bind_source_ip_error` in `simple_httpclient_test`.

bdarnell commented on .travis.yml:
```diff
@@ -47,4 +70,13 @@ install:
+      else
+         pip install -U cibuildwheel
+         cibuildwheel --output-dir dist
+         sudo rm -rf dist/*manylinux2010_x86_64.whl
```
@ploxiln I got a notification for your comment in email but github's UI has eaten it so I'm starting a new thread here.

It looks like what's going on is that `cibuildwheel` is producing two wheel files, one tagged `manylinux1` and one tagged `manylinux2010`. Then, because `cibuildwheel` doesn't seem to produce a manifest of what it built and encourages you to just do `*.whl`, there's no clean way to handle the output. 

I don't know what's going on internally with `cibuildwheel` and `manylinux` versions, but for our simple wheel it seems unlikely that we need to produce multiple versions. Can we limit it to just one manylinux version?

ploxiln replied:
It looks like we could possibly set `CIBW_MANYLINUX_X86_64_IMAGE=manylinux2010` and/or `CIBW_MANYLINUX_AARCH64_IMAGE=manylinux2014`

https://cibuildwheel.readthedocs.io/en/stable/options/#manylinux-image

> For architectures other than x86 (x86_64 and i686) manylinux2014 must be used because this is the first version of the manylinux specification that supports additional architectures.

odidev replied:
@bdarnell, yes it is creating two wheel files with two different tags(manylinux1 and manylinux2010) in x86 machine and as @ploxiln suggested we can specify `CIBW_MANYLINUX_X86_64_IMAGE=manylinux1` for x86 platform to avoid above issue.
Please check and let me know if any more changes required.

bdarnell commented:
Looks good! Thanks for all your patience with iterations on this change!

I'll follow up with getting these wheels uploaded to pypi for the 6.1 release (hopefully I'll be able to build a beta this weekend).
